src/test/regress/expected/subquery_in_targetlist.out

```diff
@@ -0,0 +1,431 @@
+FROM events_table e
+GROUP BY 1
+ORDER BY 1 LIMIT 3;
+                      u
+---------------------------------------------------------------------
+ (1,"Thu Nov 23 09:26:42.145043 2017",1,3,3,)
+
+(2 rows)
+
+-- sublink with anonymous record type
+SELECT (SELECT (user_id,value_1) FROM users_table u WHERE u.user_id = e.user_id AND time = 'Thu Nov 23 09:26:42.145043 2017')
+FROM events_table e
+WHERE user_id < 3
+GROUP BY 1
+ORDER BY 1 LIMIT 3;
+ERROR:  input of anonymous composite types is not implemented
+-- complex query using row_to_json
+SELECT coalesce(json_agg(root ORDER BY user_id), '[]') AS root
+FROM
+  (SELECT row_to_json(
+                        (SELECT _1_e
+                         FROM
+                           (SELECT "_0_root.base".user_id AS user_id) AS _1_e)) AS root, user_id
+   FROM
+     (SELECT DISTINCT user_id FROM public.users_table ORDER BY 1) AS "_0_root.base") AS _2_root ;
+                                            root
+---------------------------------------------------------------------
+ [{"user_id":1}, {"user_id":2}, {"user_id":3}, {"user_id":4}, {"user_id":5}, {"user_id":6}]
+(1 row)
+
+SELECT *
+FROM
+  (SELECT
+  	row_to_json((SELECT _1_e FROM (SELECT user_id) AS _1_e)) AS root, user_id
+   FROM
+     (SELECT DISTINCT user_id FROM public.users_table ORDER BY 1) as bar) AS foo ORDER BY user_id;
+     root      | user_id
+---------------------------------------------------------------------
+ {"user_id":1} |       1
+ {"user_id":2} |       2
+ {"user_id":3} |       3
+ {"user_id":4} |       4
+ {"user_id":5} |       5
+ {"user_id":6} |       6
+(6 rows)
+
+-- non-colocated subquery join
+SELECT count(*) FROM
+	(SELECT event_type, (SELECT e.value_2 FROM users_reference_table WHERE user_id = 1 AND value_1 = 1 AND value_2 = 1), (SELECT e.value_2)
+		FROM events_table e) as foo
+	JOIN
+	(SELECT event_type, (SELECT e.value_2 FROM users_reference_table WHERE user_id = 5 AND value_1 = 1 AND value_2 = 1), (SELECT e.value_2)
+		FROM events_table e) as bar
+	ON bar.event_type = foo.event_type;
+ count
+---------------------------------------------------------------------
+  1993
+(1 row)
+
+-- subquery in the target list in HAVING should be fine
+SELECT
+	user_id, count(*)
+FROM
+	events_table e1
+GROUP BY user_id
+	HAVING
+		count(*) > (SELECT count(*) FROM (SELECT
+					  (SELECT sum(user_id)  FROM users_table WHERE user_id = u1.user_id GROUP BY user_id)
+					FROM users_table u1
+					GROUP BY user_id) as foo) ORDER BY 1 DESC;
+ user_id | count
+---------------------------------------------------------------------
+       6 |    10
+       5 |    14
+       4 |    17
+       3 |    21
+       2 |    24
+       1 |    15
+(6 rows)
+
+-- make sure that we don't pushdown subqueries in the target list if no FROM clause
+SELECT (SELECT DISTINCT user_id FROM users_table WHERE user_id = (SELECT max(user_id) FROM users_table ));
```

Comment:
noting the summary of the discussion we had on the chat for future reference and some ideas on how to resolve:

I think this error message is an improvement for sure for the majority of the cases. But, there is no correlation in this query, or queries like:
```SQL
 SELECT (SELECT DISTINCT user_id FROM users_table u1 );
ERROR:  correlated subqueries are not supported when the FROM clause contains a subquery without FROM
```

As far as I can see, before this patch, the idea was that in order for such query to fail, it has to  be correlated, hence no recursive planner kicks in.

Now, the situation is a little different. I can think of 3 options, all works fine for me:

1. Consider avoiding `correlated ` in the error message. (Con: In majority of the cases,`correlated` is very useful keyword for the users)
2. Decide on the error message based on correlation or not. (Con: We need to write code for it)
3. Leave as-it (Con: Confusing cases like these)

Reply:
> The DeferErrorIfFromClauseRecurs check in DeferErrorIfCannotPushdownSubquery seems too strict / unnecessary and prevents nested correlated subqueries, but maybe best left to a separate PR.

Ah, forgot to mention that the root of the problem is as noted by Marco. My above notes are just to improve the error messages.